girder/api/rest.py

```diff
@@ -252,6 +252,45 @@ def setCurrentUser(user):
     cherrypy.request.girderUser = user
 
 
+def setContentDisposition(filename, disposition='attachment', setHeader=True):
+    """
+    Set the content disposition header to either inline or attachment, and
+    specify a filename that is properly escaped.
+
+    :param filename: the filename to add to the content disposition header.
+    :param disposition: either 'inline' or 'attachment'.  None is the same as
+        'attachment'.  Any other value skips setting the content disposition
+        header.
+    :param setHeader: if False, return the value that would be set to the
+        Content-Disposition header, but do not set it.
+    :returns: the content-disposition header value.
+    """
+    if disposition is None:
+        disposition = 'attachment'
+    if ((disposition not in ('inline', 'attachment') and
+         not disposition.startswith('form-data')) or not filename):
+        return
+    if not isinstance(disposition, six.binary_type):
+        disposition = disposition.encode('ascii', 'ignore')
+    if not isinstance(filename, six.text_type):
+        filename = filename.decode('utf8', 'ignore')
+    # Using ASCII for the safeFilename is more restrictive than necessary, but
+    # should work on all browsers/
+    safeFilename = filename.encode('ascii', 'ignore').replace(b'"', b'')
+    encodedFilename = filename.encode('utf8', 'ignore')
+    value = disposition + b'; filename="' + safeFilename + b'"'
+    if safeFilename != encodedFilename:
```

Comment:
[Flask's implementation of this](https://github.com/pallets/flask/blob/master/flask/helpers.py#L546) seems a lot cleaner on the surface. Do you think it would be worthwhile to crib more of what they have (potentially utilizing the `unicodedata` library)?

Reply:
Flask defers a bunch of the work to werkzeug.  In werkzeug's Header() class sending a list of values is supported, which is then encoded by werkzeug, handling unicode and string/bytes appropriately.  Multiple parameters are added in an arbitrary order (so in Flask, the `filename*` parameter might be before the `filename` parameter), which doesn't seem to affect any browser (see http://greenbytes.de/tech/tc2231/).

Flask uses the `unicodedata` library to decompose the unicode before making the fallback `filename` value, which is actually a clever way to de-accent characters where possible while converting to ISO-8859-1.

Reply:
We'd have to modify setResponseHeader to accept a werkzeug-like dictionary of parameters to make our implementation look as 'clean' as Flask's.